osquery/sql/sqlite_util.cpp

```diff
@@ -459,41 +459,115 @@ Status QueryPlanner::applyTypes(TableColumns& columns) {
   return Status(0);
 }
 
-int queryDataCallback(void* argument, int argc, char* argv[], char* column[]) {
-  if (argument == nullptr) {
-    VLOG(1) << "Query execution failed: received a bad callback argument";
-    return SQLITE_MISUSE;
+// Wrapper for legacy method until all uses can be replaced
+Status queryInternal(const std::string& query,
+                     QueryData& results,
+                     const SQLiteDBInstanceRef& instance) {
+  QueryDataTyped typedResults;
+  Status status = queryInternal(query, typedResults, instance);
+  results.reserve(typedResults.size());
+  for (const auto& row : typedResults) {
+    Row r;
+    for (const auto& col : row) {
+      r[col.first] = boost::lexical_cast<std::string>(col.second);
+    }
+    results.push_back(std::move(r));
   }
+  return status;
+}
 
-  auto qData = static_cast<QueryData*>(argument);
-  Row r;
-  for (int i = 0; i < argc; i++) {
-    if (column[i] != nullptr) {
-      if (r.count(column[i])) {
-        // Found a column name collision in the result.
-        VLOG(1) << "Detected overloaded column name " << column[i]
-                << " in query result consider using aliases";
-      }
-      r[column[i]] = (argv[i] != nullptr) ? argv[i] : FLAGS_nullvalue;
+Status readRows(sqlite3_stmt* prepared_statement,
+                QueryDataTyped& results,
+                const SQLiteDBInstanceRef& instance) {
+  int rc = sqlite3_step(prepared_statement);
+  /* if we have a result set row... */
+  if (SQLITE_ROW == rc) {
+    // First collect the column names
+    int num_columns = sqlite3_column_count(prepared_statement);
+    std::vector<std::string> colNames;
+    colNames.reserve(num_columns);
+    for (int i = 0; i < num_columns; i++) {
+      colNames.push_back(sqlite3_column_name(prepared_statement, i));
     }
+
+    do {
+      RowTyped row;
+      for (int i = 0; i < num_columns; i++) {
+        switch (sqlite3_column_type(prepared_statement, i)) {
+        case SQLITE_INTEGER:
+          row[colNames[i]] =
+              static_cast<int64_t>(sqlite3_column_int64(prepared_statement, i));
+          break;
+        case SQLITE_FLOAT:
+          row[colNames[i]] = sqlite3_column_double(prepared_statement, i);
+          break;
+        case SQLITE_NULL:
+          row[colNames[i]] = FLAGS_nullvalue;
+          break;
+        default:
+          // Everything else (SQLITE_TEXT, SQLITE3_TEXT, SQLITE_BLOB) is
+          // obtained/conveyed as text/string
+          row[colNames[i]] = std::string(reinterpret_cast<const char*>(
+              sqlite3_column_text(prepared_statement, i)));
+        }
+      }
+      results.push_back(std::move(row));
+      rc = sqlite3_step(prepared_statement);
+    } while (SQLITE_ROW == rc);
+  }
+  if (!(rc == SQLITE_OK || rc == SQLITE_DONE)) {
```

Comment:
I think you should always get here with a `SQLITE_DONE` code and `sqlite3_step()` doesn't return `SQLITE_OK`. But since the documentation isn't very clear please test it for the edge case of no results.